File: codes/julia/ch19/semi_infinite_compare.jl
# semi_infinite_compare.jl
# Chapter 19: Coordinate Transformations
# Computational Etude 19.4: Infinity without truncation.
#
# Author: Dr. Denys Dutykh (Khalifa University, Abu Dhabi, UAE)

using CairoMakie
using LinearAlgebra
using Printf

include(joinpath(@__DIR__, "..", "ch07", "cheb_matrix.jl"))

function solve_truncation(N, L)
    D, xi = cheb_matrix(N)
    y = 0.5 * L .* (xi .+ 1)
    Dy = (2 / L) .* D
    Dy2 = Dy * Dy
    A = Dy2[2:N, 2:N] - I(N - 1)
    rhs = -Dy2[2:N, 1] .* 0.0 .- Dy2[2:N, N + 1] .* 1.0
    u_int = A \ rhs
    u = zeros(N + 1); u[1] = 0.0; u[N + 1] = 1.0; u[2:N] = u_int
    return y, u
end

function solve_algebraic(N, ell)
    D, x = cheb_matrix(N)
    fp  = @. 2ell / (1 - x)^2
    fpp = @. 4ell / (1 - x)^3
    Dy = Diagonal(1 ./ fp) * D
    Dy2 = Diagonal(1 ./ fp .^ 2) * (D * D) - Diagonal(fpp ./ fp .^ 3) * D
    y = @. ell * (1 + x) / (1 - x)
    A = Dy2[2:N, 2:N] - I(N - 1)
    rhs = -Dy2[2:N, 1] .* 0.0 .- Dy2[2:N, N + 1] .* 1.0
    u_int = A \ rhs
    u = zeros(N + 1); u[1] = 0.0; u[N + 1] = 1.0; u[2:N] = u_int
    return y, u
end

function max_err(y, u)
    mask = isfinite.(y) .& (y .< 1e6)
    return maximum(abs.(u[mask] .- exp.(-y[mask])))
end

function run()
    outdir = joinpath(@__DIR__, "..", "..", "..",
                      "textbook", "figures", "ch19", "julia")
    mkpath(outdir)
    Ns = [12, 16, 20, 24, 32, 40, 48, 64]
    L_trunc = [10, 20, 40]   # truncation half-widths
    ell_map = [1, 2, 4, 8]   # map parameters
    E_trunc = zeros(length(Ns), length(L_trunc))
    E_map   = zeros(length(Ns), length(ell_map))
    for (i, N) in enumerate(Ns)
        for (j, L) in enumerate(L_trunc)
            y, u = solve_truncation(N, L); E_trunc[i, j] = maximum(abs.(u .- exp.(-y)))
        end
        for (j, ell) in enumerate(ell_map)
            y, u = solve_algebraic(N, ell); E_map[i, j] = max_err(y, u)
        end
    end

    NAVY = colorant"#142D6E"; CORAL = colorant"#E74C3C"; TEAL = colorant"#16A085"
    ORANGE = colorant"#E67E22"; GOLD = colorant"#D4A017"; PURPLE = colorant"#8E44AD"
    fig = Figure(size=(1300, 340))

    ax1 = Axis(fig[1, 1], xlabel="y", ylabel="u(y)", title="(a) Solution at N=24")
    y_t, u_t = solve_truncation(24, 20); y_m, u_m = solve_algebraic(24, 2)
    yplot = collect(range(0, 12, length=401))
    lines!(ax1, yplot, exp.(-yplot); color=NAVY, linewidth=1.2)
    scatter!(ax1, y_t, u_t; color=CORAL, markersize=5)
    y_m_clip = clamp.(y_m, 0, 12)
    scatter!(ax1, y_m_clip, u_m; color=TEAL, markersize=5)

    ax2 = Axis(fig[1, 2], xlabel="N", ylabel="max error", yscale=log10,
               title="(b) Truncation")
    colours_t = [CORAL, ORANGE, GOLD]
    for (j, (L, c)) in enumerate(zip(L_trunc, colours_t))
        scatterlines!(ax2, Ns, E_trunc[:, j]; color=c, label="L=$L")
    end
    axislegend(ax2; position=:rt)

    ax3 = Axis(fig[1, 3], xlabel="N", ylabel="max error", yscale=log10,
               title="(c) Algebraic map")
    colours_m = [CORAL, TEAL, PURPLE, NAVY]
    for (j, (ell, c)) in enumerate(zip(ell_map, colours_m))
        scatterlines!(ax3, Ns, max.(E_map[:, j], 1e-18); color=c, label="ℓ=$ell")
    end
    axislegend(ax3; position=:rt)

    save(joinpath(outdir, "semi_infinite_compare.pdf"), fig)
    save(joinpath(outdir, "semi_infinite_compare.png"), fig)
    @printf("[19.4-julia] saved figure\n")
end

if abspath(PROGRAM_FILE) == @__FILE__
    run()
end

File: codes/matlab/ch19/semi_infinite_compare.m
function semi_infinite_compare()
%% semi_infinite_compare - Chapter 19, Computational Etude 19.4.
%
% Semi-infinite BVP  u'' - u = 0,  u(0)=1, u(infty)=0,  with exact
% solution u_ex(y) = exp(-y).  Two methods:
%   (I)  truncation to [0, L];
%   (II) algebraic map  y = L (1+x)/(1-x).
%
% Author: Dr. Denys Dutykh (Khalifa University, Abu Dhabi, UAE)

    script_dir = fileparts(mfilename('fullpath'));
    addpath(fullfile(script_dir, '..', 'ch07'));
    out_dir = fullfile(script_dir, '..', '..', '..', ...
                       'textbook', 'figures', 'ch19', 'matlab');
    if ~exist(out_dir, 'dir'); mkdir(out_dir); end

    Ns = [12 16 20 24 32 40 48 64];
    L_trunc = [10 20 40];     % truncation half-widths
    ell_map = [1 2 4 8];      % map parameters
    err_trunc = zeros(length(Ns), length(L_trunc));
    err_map   = zeros(length(Ns), length(ell_map));
    for i = 1:length(Ns)
        N = Ns(i);
        for j = 1:length(L_trunc)
            [y, u] = solve_truncation(N, L_trunc(j));
            err_trunc(i,j) = max(abs(u - exp(-y)));
        end
        for j = 1:length(ell_map)
            [y, u] = solve_algebraic(N, ell_map(j));
            mask = isfinite(y) & y < 1e6;
            err_map(i,j) = max(abs(u(mask) - exp(-y(mask))));
        end
    end

    NAVY=[20 45 110]/255; CORAL=[231 76 60]/255; TEAL=[22 160 133]/255;
    ORANGE=[230 126 34]/255; PURPLE=[142 68 173]/255; GOLD=[212 160 23]/255;

    Y_LIM = 12.0;
    Ngrid = 24;
    [~, x_demo] = cheb_matrix(Ngrid);
    x_line = linspace(-1.0, 1.0 - 1e-12, 401)';

    fig = figure('Position',[100 100 1100 760],'Color','w');

    %% (a) Solution at N=24, both methods overlaid on exact
    subplot(2,2,1);
    [yt, ut] = solve_truncation(24, 20);
    [ym, um] = solve_algebraic(24, 2);
    yplot = linspace(0, 12, 401);
    plot(yplot, exp(-yplot), '-', 'Color', NAVY, 'LineWidth', 1.4); hold on;
    plot(yt, ut, 'o', 'Color', CORAL, 'MarkerSize', 5);
    ym_c = min(ym, 12);
    plot(ym_c, um, 's', 'Color', TEAL, 'MarkerSize', 5);
    xlim([0 12]); ylim([-0.05 1.1]); grid on; box on;
    xlabel('$y$', 'Interpreter','latex');
    ylabel('$u(y)$', 'Interpreter','latex');
    title('(a) Solution at $N = 24$', 'Interpreter','latex');
    legend({'exact', 'truncation $L = 20$', 'algebraic $\ell = 2$'}, ...
        'Interpreter','latex', 'Location', 'northeast', 'Box', 'off');

    %% (b) NEW: algebraic map shape y(x) for the four ell values
    subplot(2,2,2); hold on;
    colours_map = [CORAL; TEAL; PURPLE; NAVY];
    legendH = gobjects(1, length(ell_map));
    for j = 1:length(ell_map)
        ell = ell_map(j);
        y_curve = ell*(1 + x_line)./(1 - x_line);
        visible = y_curve <= Y_LIM;
        legendH(j) = plot(x_line(visible), y_curve(visible), ...
            '-', 'Color', colours_map(j,:), 'LineWidth', 1.6);
    end
    plot(x_demo, -0.7*ones(size(x_demo)), '|', ...
         'Color', NAVY, 'MarkerSize', 12);
    text(-1.0, -1.1, 'CGL nodes $x_j$', 'Interpreter','latex', ...
         'FontSize', 8, 'Color', NAVY);
    plot([-1.10 1.10], [0 0], 'Color', [0.85 0.85 0.85], 'LineWidth', 0.4);
    plot([0 0], [-1.6 Y_LIM+1.0], 'Color', [0.85 0.85 0.85], 'LineWidth', 0.4);
    xlim([-1.10 1.10]); ylim([-1.6 Y_LIM + 1.0]); box on;
    xlabel('computational coordinate $x$', 'Interpreter','latex');
    ylabel('physical $y$', 'Interpreter','latex');
    title('(b) Algebraic map $y = \ell(1+x)/(1-x)$', 'Interpreter','latex');
    legend(legendH, ...
        arrayfun(@(e) sprintf('$\\ell = %d$', e), ell_map, ...
                 'UniformOutput', false), ...
        'Interpreter','latex', 'Location', 'northwest', 'Box', 'off');

    %% (c) Truncation: error vs N
    subplot(2,2,3);
    colours = [CORAL; ORANGE; GOLD]; hold on;
    for j = 1:length(L_trunc)
        semilogy(Ns, err_trunc(:,j), '-o', 'Color', colours(j,:), ...
                 'LineWidth', 1.1);
    end
    set(gca, 'YScale', 'log'); grid on; box on;
    xlabel('$N$', 'Interpreter','latex');
    ylabel('$\|u - u_N\|_\infty$', 'Interpreter','latex');
    title('(c) Truncation: error vs $N$', 'Interpreter','latex');
    legend(arrayfun(@(L) sprintf('trunc.\\ $L = %d$', L), L_trunc, ...
                    'UniformOutput', false), ...
           'Interpreter','latex', 'Location', 'best');

    %% (d) Algebraic map: error vs N
    subplot(2,2,4);
    colours = [CORAL; TEAL; PURPLE; NAVY]; hold on;
    for j = 1:length(ell_map)
        semilogy(Ns, err_map(:,j), '-s', 'Color', colours(j,:), ...
                 'LineWidth', 1.1, 'MarkerFaceColor', 'none');
    end
    set(gca, 'YScale', 'log'); grid on; box on;
    xlabel('$N$', 'Interpreter','latex');
    ylabel('$\|u - u_N\|_\infty$', 'Interpreter','latex');
    title('(d) Algebraic map: error vs $N$', 'Interpreter','latex');
    legend(arrayfun(@(e) sprintf('algebraic $\\ell = %d$', e), ell_map, ...
                    'UniformOutput', false), ...
           'Interpreter','latex', 'Location', 'best');

    set(fig, 'PaperPositionMode','auto');
    pos = get(fig, 'Position');
    set(fig, 'PaperUnits','points', ...
             'PaperSize',[pos(3) pos(4)], ...
             'PaperPosition',[0 0 pos(3) pos(4)]);
    print(fig, fullfile(out_dir, 'semi_infinite_compare.pdf'), '-dpdf');
    print(fig, fullfile(out_dir, 'semi_infinite_compare.png'), '-dpng');
    close(fig);
    fprintf('[19.4-matlab] figure saved\n');
end

function [y, u] = solve_truncation(N, L)
    [D, x] = cheb_matrix(N);
    y = 0.5*L*(x+1);
    Dy = (2/L)*D;  Dy2 = Dy*Dy;
    A = Dy2(2:N, 2:N) - eye(N-1);
    rhs = -Dy2(2:N, 1)*0 - Dy2(2:N, N+1)*1;
    u_int = A \ rhs;
    u = zeros(N+1,1); u(1) = 0; u(N+1) = 1; u(2:N) = u_int;
end

function [y, u] = solve_algebraic(N, ell)
    [D, x] = cheb_matrix(N);
    fp  = 2*ell ./ (1 - x).^2;
    fpp = 4*ell ./ (1 - x).^3;
    Dy  = diag(1./fp)*D;
    Dy2 = diag(1./fp.^2)*(D*D) - diag(fpp./fp.^3)*D;
    y = ell*(1+x)./(1-x);
    A = Dy2(2:N, 2:N) - eye(N-1);
    rhs = -Dy2(2:N, 1)*0 - Dy2(2:N, N+1)*1;
    u_int = A \ rhs;
    u = zeros(N+1,1); u(1) = 0; u(N+1) = 1; u(2:N) = u_int;
end
